### PR TITLE
Support coverage across multiple test packages.

### PR DESCRIPTION
**Pull request by cmars**
Reworked #65.

### CODE REVIEW

**Comment by axw** [comment]
This is looking better, thanks, but there is a problem: with this, you would no longer be able to pass additional build flags or arguments to the test program.

**Comment by cmars** [comment]
Good point. PTAL.

File: gocov/test.go [hunk 1]
```diff
@@ -21,27 +21,93 @@
...
+			resolvedPkgs = append(resolvedPkgs, line)
+		}
+	}
+	return resolvedPkgs, nil
+}
+
+func splitPkgsFlags(args []string) ([]string, []string) {
+	var pkgs, flags []string
+	for _, arg := range args {
+		if len(arg) > 0 && arg[0] == '-' {
```
Comment (axw) [inline gocov/test.go]: getting into pedantic territory, but this should handle "--" too:

```
for i, arg := range args {
    if arg == "--" {
        flags = append(flags, args[i:]...)
        break
    }
    ...
}
```

Comment (axw) [inline gocov/test.go]: This does mean you *have* to use "-flag=value" and never "-flag value", and if you don't you're going to get weird errors... but I think that's fine given the alternatives.

**Comment by axw** [comment]
LGTM with -- handled. Thank you.

**Comment by cmars** [comment]
`go test` seems to treat everything as a flag or flag arguments after encountering the first flag. By adopting the same behavior, this fixes space-delimited flag arguments, and I don't think there's a need for a `--`.

**Comment by axw** [comment]
> go test seems to treat everything as a flag or flag arguments after encountering the first flag.

Nope. https://github.com/golang/go/blob/master/src/cmd/go/testflag.go#L82

But that's okay, I'm not fussed about supporting either/or. I think I'd only be completely satisfied if we handled all flags the same as go test, but I think users can deal with this.

**Comment by mshean** [comment]
This change broke all of our builds which rely on gocov. We implemented coverage across multiple packages with the below code statement:

```
        for dir in $$(find . -not -path './.git*' -not -path '*/_*' -type d); do \
                if ls $$dir/*.go &> /dev/null; then \
                        xmlname=`echo $$dir | sed -e 's/\.\///g; s/\//_/g'`; \
                        $(GOPATH)/bin/gocov test -v $$dir | $(GOPATH)/bin/gocov-xml > $(COVERAGE_DIR)/$$xmlname.xml; \
                fi; \
        done; \
```

Note the important line `if ls $$dir/*.go`, we only want to run gocov test if there are actual go files to test in the directory. Running gocov test in a directory which is missing go files causes gocov-xml to panic.

After this change we are getting these errors (actual file names redacted):

```
18:52:15 goroutine 16 [running]:
18:52:15 runtime.panic(0x515520, 0xc208000090)
18:52:15 	/usr/local/bin/go/src/pkg/runtime/panic.c:279 +0xf5
18:52:15 main.main()
18:52:15 	/var/builds/workspace/_build/src/github.com/AlekSi/gocov-xml/gocov-xml.go:60 +0x130
18:52:15 
18:52:15 goroutine 19 [finalizer wait]:
18:52:15 runtime.park(0x414170, 0x600ef8, 0x5ffa09)
18:52:15 	/usr/local/bin/go/src/pkg/runtime/proc.c:1369 +0x89
18:52:15 runtime.parkunlock(0x600ef8, 0x5ffa09)
18:52:15 	/usr/local/bin/go/src/pkg/runtime/proc.c:1385 +0x3b
18:52:15 runfinq()
18:52:15 	/usr/local/bin/go/src/pkg/runtime/mgc0.c:2644 +0xcf
18:52:15 runtime.goexit()
18:52:15 	/usr/local/bin/go/src/pkg/runtime/proc.c:1445foobar
18:52:15 can't load package: package foobar: no buildable Go source files in /var/builds/workspace/_build/src/foobar
18:52:15 error: exit status 1
18:52:15 panic: EOF
18:52:15 
18:52:15 goroutine 16 [running]:
18:52:15 runtime.panic(0x515520, 0xc208000090)
18:52:15 	/usr/local/bin/go/src/pkg/runtime/panic.c:279 +0xf5
18:52:15 main.main()
18:52:15 	/var/builds/workspace/_build/src/github.com/AlekSi/gocov-xml/gocov-xml.go:60 +0x130
18:52:15 
18:52:15 goroutine 19 [finalizer wait]:
18:52:15 runtime.park(0x414170, 0x600ef8, 0x5ffa09)
18:52:15 	/usr/local/bin/go/src/pkg/runtime/proc.c:1369 +0x89
18:52:15 runtime.parkunlock(0x600ef8, 0x5ffa09)
18:52:15 	/usr/local/bin/go/src/pkg/runtime/proc.c:1385 +0x3b
18:52:15 runfinq()
18:52:15 	/usr/local/bin/go/src/pkg/runtime/mgc0.c:2644 +0xcf
18:52:15 runtime.goexit()
18:52:15 	/usr/local/bin/go/src/pkg/runtime/proc.c:1445
18:52:15 can't load package: package foobar: no buildable Go source files in /var/builds/workspace/_build/src/foobar
18:52:15 error: exit status 1
18:52:15 panic: EOF
```

directory structure looks like this

foobar/GLOCKFILE
foobar/package1/
foobar/package1/main.go
foobar/package1/abcd/abcd.go
foobar/package1/defg/defg.go
foobar/package1/bin/something.sh
foobar/package1/conf/something.conf

**Comment by mshean** [comment]
Specifying a directory does not seem to work anymore (e.g. gocov test src/test/foobar), it will run gocov test in ./ instead of src/test/foobar.

To test multiple multiple packages, we used this loop:
```
        for dir in $$(find . -not -path './.git*' -not -path '*/_*' -type d); do \
                if ls $$dir/*_test.go &> /dev/null; then \
                        xmlname=`echo $$dir | sed -e 's/\.\///g; s/\//_/g'`; \
                        $(GOPATH)/bin/gocov test -v $$dir | $(GOPATH)/bin/gocov-xml > $(COVERAGE_DIR)/$$xmlname.xml; \
                fi; \
        done; \
```

I had to change it to this to avoid the gocov-xml panic errors:
```
        for dir in $$(find . -not -path './.git*' -not -path '*/_*' -type d); do \
                if ls $$dir/*_test.go &> /dev/null; then \
                        cd $$dir; \
                        xmlname=`echo $$dir | sed -e 's/\.\///g; s/\//_/g'`; \
                        $(GOPATH)/bin/gocov test -v | $(GOPATH)/bin/gocov-xml > $(COVERAGE_DIR)/$$xmlname.xml; \
                        cd - >/dev/null; \
                fi; \
        done; \
```

**Comment by axw** [comment]
@mshean Ugh, that's a silly oversight. Sorry. I'll come up with something.

**Comment by axw** [comment]
@mshean You might not care any more since you worked around it, but this should be resolved in 7680dd3a4cdcd085af87cad319460b289da30024. I've adapted the code from cmd/go that splits flags, so we should now be entirely compatible (excluding flags that don't make sense to handle).

**Comment by mshean** [comment]
@axw awesome thanks for the quick fix

**Pull request merged**